File: 1/2021.03.28/222.Count Complete Tree Nodes/python/solution.py
class Solution:
    def countNodes(self, root: TreeNode) -> int:
        l = r = root
        hl = hr = 0
        while l:
            l = l.left
            hl += 1
        while r:
            r = r.right
            hr += 1
        if hl == hr:
            return pow(2, hl) - 1
        return 1 + self.countNodes(root.left) + self.countNodes(root.right)

File: 1/2021.03.28/222.Count Complete Tree Nodes/java/solution.java
class Solution {
    public int countNodes(TreeNode root) {
        TreeNode l = root;
        TreeNode r = root;
        int hl = 0;
        int hr = 0;
        while (l != null) {
            l = l.left;
            hl++;
        }
        while (r != null) {
            r = r.right;
            hr++;
        }
        if (hl == hr) {
            return (int)Math.pow(2, hl) - 1;
        }
        return 1 + countNodes(root.left) + countNodes(root.right);
    }
}
